crates/re_query/src/range/query.rs

```diff
@@ -47,7 +48,65 @@ impl Caches {
             };
 
             let mut cache = cache.write();
+
+            // TODO(#4810): Get rid of this once we have proper bucketing in place.
+            //
+            // Detects the case where the user loads a piece of data at the end of the time range, then a piece
+            // at the beginning of the range, and finally a piece right in the middle.
+            //
+            // DATA = ###################################################
+            //          |      |     |       |            \_____/
+            //          \______/     |       |            query #1
+            //          query #2     \_______/
+            //                       query #3
+            //
+            // and coarsly invalidates the whole range in that case, to avoid the kind of bugs
+            // showcased in <https://github.com/rerun-io/rerun/issues/5686>.
+            {
+                let time_range = cache.per_data_time.read_recursive().time_range();
+                if let Some(time_range) = time_range {
+                    {
+                        let hole_start = time_range.max();
+                        let hole_end =
+                            TimeInt::new_temporal(query.range().min().as_i64().saturating_sub(1));
+                        if hole_start < hole_end {
```

Comment:
> It looks like what's going on is that is that any time a new query comes before or after the existing range cache, then we clear the cache to avoid ever ending up with a hole in the first place?

Yes, we prematurely clear the cache anytime two non-contiguous queries (as defined by the checks below) come in one after another, to keep it simple.